test/e2e/storage/api/testsuite.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	"fmt"
+
+	"github.com/onsi/ginkgo"
+
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	e2evolume "k8s.io/kubernetes/test/e2e/framework/volume"
+)
+
+// TestSuite represents an interface for a set of tests which works with TestDriver.
+// Each testsuite should implement this interface and use a TestSuiteHandler to define its Tests.
+// RegisterTests function should be called to actually register tests for the TestSuite.
+// All the functions except GetTestSuiteInfo are designed to not be exported so that caller
+// has to use RegisterTests() to register the tests in a standard way.
+type TestSuite interface {
+	GetTestSuiteInfo() TestSuiteInfo
+	// defineTests defines tests of the testpattern for the driver.
+	// Called inside a Ginkgo context that reflects the current driver and test pattern,
+	// so the test suite can define tests directly with ginkgo.It.
+	defineTests(TestDriver, TestPattern)
+	// skipUnsupportedTests will skip the test suite based on the given TestPattern, TestDriver
+	// Testsuite should check if the given pattern and driver works for the "whole testsuite"
+	// Testcase specific check should happen inside defineTests
+	skipUnsupportedTests(TestDriver, TestPattern)
+}
+
+// RegisterTests register the driver + pattern combination to the inside TestSuite
+// This function actually register tests inside testsuite
+func RegisterTests(suite TestSuite, driver TestDriver, pattern TestPattern) {
+	tsInfo := suite.GetTestSuiteInfo()
+	testName := fmt.Sprintf("[Testpattern: %s]%s %s%s", pattern.Name, pattern.FeatureTag, tsInfo.Name, tsInfo.FeatureTag)
+	ginkgo.Context(testName, func() {
+		ginkgo.BeforeEach(func() {
+			// skip all the invalid combination of driver and pattern first
+			SkipInvalidDriverPatternCombination(driver, pattern)
+			// skip the invalid test pattern and driver combination
+			suite.skipUnsupportedTests(driver, pattern)
```

Comment:
what is the difference between these two skips?

Reply:
The first is globally we filter out some invalid combination. So something like if the pattern ask for inlineVolume but the driver does not support inline at all. It will skip.

The second skip is a testsuite specific skip. Each testsuite will have some certain test patterns or drivers that it does not support to run. So we can add those there.